File: onecycle/onecycle.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	ifstream in("input.in");
	ofstream out("output.out");

	auto my_assert = [&](bool cond, const char* comment) {
		if (!cond) {
			out << "Assertion Failed: " << comment << '\n';
			exit(0);
		}
	};
	
	int graph_size; in >> graph_size;
	vector<vector<int>> adj(graph_size);
	for (int i = 0; i < graph_size; i++) {
		int u, v;
		in >> u >> v;
		--u, --v;
		my_assert(u != v && 0 <= u && u < graph_size && 0 <= v && v < graph_size, "wrong input");
		adj[u].push_back(v);
		adj[v].push_back(u);
	}
	vector<int> ecc(graph_size);

	vector<bool> cyc_visited(graph_size);
	vector<int> parent(graph_size);
	int backstart = -1, backend = -1;
	function<void(int, int)> cyc_dfs = [&](int from, int pre) {
		parent[from] = pre;
		cyc_visited[from] = true;
		for (int to : adj[from]) {
			if (to != pre) {
				if (cyc_visited[to]) {
					backstart = from, backend = to;
					return;
				}
				else {
					cyc_dfs(to, from);
					if (backstart != -1) {
						return;
					}
				}
			}
		}
	};
	cyc_dfs(0, -1);
	my_assert(backstart != -1, "no cycles");
	
	vector<int> cyc;
	int backcur = backstart;
	while (backcur != backend) {
		cyc.push_back(backcur);
		backcur = parent[backcur];
	}
	cyc.push_back(backend);

	int cyc_size = cyc.size();
    vector<vector<int>> trees(cyc_size);
	vector<int> depth(graph_size, -1), max_depth(cyc_size, -1);
	vector<int> dist(graph_size, -1);
	for (int i = 0; i < cyc_size; i++) {
		int root = cyc[i];
		// ecc[root] = max(ecc[root], cyc_size / 2);
		function<void(int, int, vector<int>&)> tree_dfs = [&](int from, int pre, vector<int>& d) {
			trees[i].push_back(from);
			for (int to : adj[from]) {
				if (to != pre && to != cyc[(i - 1 + cyc_size) % cyc_size] && to != cyc[(i + 1) % cyc_size]) {
					d[to] = d[from] + 1;
					tree_dfs(to, from, d);
				}
			}
		};
		depth[root] = 0;
		tree_dfs(root, -1, depth);

		int deepest = -1;
		for (int v : trees[i]) {
			if (max_depth[i] < depth[v]) {
				max_depth[i] = depth[v];
				deepest = v;
			}
		}
		my_assert(deepest != -1, "depth array is incorrect");

		dist[deepest] = 0;
		tree_dfs(deepest, -1, dist);
		int diam_len = -1, farthest = -1;
		for (int v : trees[i]) {
			if (diam_len < dist[v]) {
				diam_len = dist[v];
				farthest = v;
			}
		}
		my_assert(farthest != -1, "dist array is incorrect");

		for (int v : trees[i]) {
			ecc[v] = max(ecc[v], dist[v]);
		}
		dist[farthest] = 0;
		tree_dfs(farthest, -1, dist);
		for (int v : trees[i]) {
			ecc[v] = max(ecc[v], dist[v]);
		}
	}
	
    vector<int> on_cyc(graph_size);
    auto do_half = [&]() {
        vector<int> mod_max(cyc_size);
        for (int i = 0; i < cyc_size; i++) {
            mod_max[i] = max_depth[i] + i;
        }
        vector<int> ext_cyc = mod_max;
        for (int i = 0; i < cyc_size; i++) {
            ext_cyc.push_back(mod_max[i] + cyc_size);
        }
        deque<pair<int, int>> dq;
        int interval = cyc_size / 2 + 1;
        for (int i = 0; i <= cyc_size + interval - 2; i++) {
            while (!dq.empty() && dq.front().second <= i - interval) {
                dq.pop_front();
            }
            while (!dq.empty() && dq.back().first < ext_cyc[i]) {
                dq.pop_back();
            }
            dq.emplace_back(ext_cyc[i], i);

            if (i >= interval - 1) {
                on_cyc[cyc[i - interval + 1]] = max(on_cyc[cyc[i - interval + 1]], dq.front().first - (i - interval + 1));
            }
        }
    };
    do_half();
    reverse(cyc.begin(), cyc.end());
    reverse(max_depth.begin(), max_depth.end());
    do_half();
    reverse(cyc.begin(), cyc.end());
    reverse(max_depth.begin(), max_depth.end());

    for (int i = 0; i < cyc_size; i++) {
        for (int v : trees[i]) {
            ecc[v] = max(ecc[v], on_cyc[cyc[i]] + depth[v]);
        }
    }

	for (int i = 0; i < graph_size; i++) {
        my_assert(0 <= ecc[i] && ecc[i] < graph_size, "invalid eccentricity");
		out << i << ' ' << ecc[i] << '\n';
	}
	return 0;
}

File: onecycle/onecycle.py
import sys, os
from collections import deque

sys.setrecursionlimit(10 ** 6)
__location__ = os.path.realpath(os.path.join(os.getcwd(), os.path.dirname(__file__)))
inf = open(os.path.join(__location__, "input.in"), "r")
ouf = open(os.path.join(__location__, "output.out"), "w")

def my_assert(cond, comment):
    if not cond:
        ouf.write(f"Assertion Failed: {comment}")
        ouf.close()
        sys.exit(0)

graph_size = int(inf.readline().rstrip())
adj = [[] for _ in range(graph_size)]
for i in range(graph_size):
    u, v = map(int, inf.readline().rstrip().split())
    u -= 1; v -= 1
    my_assert(u != v and 0 <= u < graph_size and 0 <= v < graph_size, "wrong input")
    adj[u].append(v)
    adj[v].append(u)
inf.close()
ecc = [0] * graph_size

cyc_visited = [False] * graph_size
parent = [0] * graph_size
backstart, backend = -1, -1
def cyc_dfs(cur, pre):
    global backstart, backend
    parent[cur] = pre
    cyc_visited[cur] = True
    for next in adj[cur]:
        if next != pre:
            if cyc_visited[next]:
                backstart, backend = cur, next
                return
            else:
                cyc_dfs(next, cur)
                if backstart != -1:
                    return
cyc_dfs(0, -1)
my_assert(backstart != -1, "no cycles")

cyc = []
backcur = backstart
while backcur != backend:
    cyc.append(backcur)
    backcur = parent[backcur]
cyc.append(backend)

cyc_size = len(cyc)
trees = [[] for _ in range(cyc_size)]
depth, max_depth = [-1] * graph_size, [-1] * cyc_size
dist = [-1] * graph_size
for i in range(cyc_size):
    root = cyc[i]
    # ecc[root] = max(ecc[root], cyc_size // 2)
    def tree_dfs(cur, pre, d, push):
        if push:
            trees[i].append(cur)
        for next in adj[cur]:
            if next not in [pre, cyc[i - 1], cyc[(i + 1) % cyc_size]]:
                d[next] = d[cur] + 1
                tree_dfs(next, cur, d, push)
    depth[root] = 0
    tree_dfs(root, -1, depth, True)
    
    deepest = -1
    for v in trees[i]:
        if max_depth[i] < depth[v]:
            max_depth[i] = depth[v]
            deepest = v
    my_assert(deepest != -1, "depth array is incorrect")

    dist[deepest] = 0
    tree_dfs(deepest, -1, dist, False)
    diam_len, farthest = -1, -1
    for v in trees[i]:
        if diam_len < dist[v]:
            diam_len = dist[v]
            farthest = v
    my_assert(farthest != -1, "dist array is incorrect")

    for v in trees[i]:
        ecc[v] = max(ecc[v], dist[v])
    dist[farthest] = 0
    tree_dfs(farthest, -1, dist, False)
    for v in trees[i]:
        ecc[v] = max(ecc[v], dist[v])
my_assert(-1 not in max_depth, "unvisited vertex exists")

on_cyc = [0] * graph_size
def do_half():
    mod_max = [max_depth[i] + i for i in range(cyc_size)]
    ext_cyc = mod_max[:]
    for i in range(cyc_size):
        ext_cyc.append(mod_max[i] + cyc_size)
    dq = deque()
    interval = cyc_size // 2
    for i in range(cyc_size + interval):
        while dq and dq[0][1] <= i - interval:
            dq.popleft()
        while dq and dq[-1][0] < ext_cyc[i]:
            dq.pop()
        dq.append((ext_cyc[i], i))

        if i >= interval:
            on_cyc[cyc[i - interval]] = max(on_cyc[cyc[i - interval]], dq[0][0] - (i - interval))
do_half()
cyc.reverse()
max_depth.reverse()
do_half()
cyc.reverse()
max_depth.reverse()

for i in range(cyc_size):
    for v in trees[i]:
        ecc[v] = max(ecc[v], on_cyc[cyc[i]] + depth[v])

cnt = [0] * graph_size
for i in range(graph_size):
    my_assert(0 <= ecc[i] < graph_size, "invalid eccentricity")
    cnt[ecc[i]] += 1
eccseq = []
for i in range(graph_size):
    for j in range(cnt[i]):
        eccseq.append(i)
for i in range(len(eccseq)):
    ouf.write(f"{eccseq[i]} ")
ouf.write("\n")
ouf.close()